File: crates/prompt/src/mentions.rs
//! # Inline `@file` mention expansion
//!
//! Web 与 CLI 共享同一套语法与渲染格式，保证两端「在发送前把 `@file`-提及展开为
//! 上下文块」行为一致（Roo-Code `parseMentions` 的精简子集）：
//!
//! - `@file <path>`（路径不含空白）→ 注入该文件内容。
//!
//! 渲染为附加在用户文本后的「上下文块」，CLI 直接调用本模块；Web（TS）按相同格式镜像。
//!
//! 格式示例：
//! ```text
//! <file path="src/auth.rs">
//! ```rust
//! …contents…
//! ```
//! </file>
//! ```

#![allow(clippy::module_name_repetitions)]

/// 一个解析出的提及。
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Mention {
    /// 文件提及（相对工作区根的路径，原样保留）。
    File(String),
}

/// 从文本解析所有 `@file <path>` 提及（去重，保持首次出现顺序）。
///
/// 语法：按空白分词；`@file` 后跟随一个不含空白的路径 token，并剥离其首尾的 `"`/`'`。
/// 大小写不敏感（`@FILE` 同样识别）。未识别的 `@xxx` 忽略。
#[must_use]
pub fn parse_mentions(text: &str) -> Vec<Mention> {
    let mut out: Vec<Mention> = Vec::new();
    for line in text.lines() {
        let mut tokens = line.split_whitespace();
        while let Some(tok) = tokens.next() {
            if tok.eq_ignore_ascii_case("@file") {
                if let Some(p) = tokens.next() {
                    let cleaned = p.trim_matches(|c: char| c == '"' || c == '\'');
                    let m = Mention::File(cleaned.to_string());
                    if !out.contains(&m) {
                        out.push(m);
                    }
                }
            }
        }
    }
    out
}

/// 按扩展名给出 Markdown 围栏语言（用于代码块高亮）。
#[must_use]
pub fn fence_lang(path: &str) -> &'static str {
    let ext = path.rsplit('.').next().unwrap_or("");
    match ext.to_ascii_lowercase().as_str() {
        "rs" => "rust",
        "toml" => "toml",
        "ts" | "tsx" => "ts",
        "js" | "jsx" | "mjs" | "cjs" => "js",
        "py" => "python",
        "go" => "go",
        "java" => "java",
        "c" | "h" => "c",
        "cpp" | "cc" | "hpp" => "cpp",
        "rb" => "ruby",
        "sh" | "bash" | "zsh" => "bash",
        "json" => "json",
        "yaml" | "yml" => "yaml",
        "md" => "md",
        "html" => "html",
        "css" => "css",
        _ => "",
    }
}

/// 渲染单个文件上下文块。
#[must_use]
pub fn format_file_block(path: &str, content: &str) -> String {
    format!(
        "<file path=\"{path}\">\n```{lang}\n{content}\n```\n</file>",
        lang = fence_lang(path),
    )
}

/// 把若干已渲染的上下文块附加到用户文本后（无块时原样返回）。
#[must_use]
pub fn render_attached(user_text: &str, blocks: &[String]) -> String {
    if blocks.is_empty() {
        return user_text.to_string();
    }
    format!(
        "{user_text}\n\n--- attached context ---\n\n{joined}",
        joined = blocks.join("\n\n"),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_file_mentions() {
        let m = parse_mentions("fix @file src/auth.rs and also @FILE x.py");
        assert_eq!(
            m,
            vec![
                Mention::File("src/auth.rs".into()),
                Mention::File("x.py".into()),
            ]
        );
    }

    #[test]
    fn dedups_repeats_and_ignores_unknown() {
        let m = parse_mentions("@unknown @file a.rs @file a.rs");
        assert_eq!(m, vec![Mention::File("a.rs".into())]);
    }

    #[test]
    fn strips_quotes_around_paths() {
        let m = parse_mentions("@file \"src/a b.rs\"");
        // 路径含空白会被 split_whitespace 切断；此处验证引号剥离（不含空白的引号路径）。
        assert_eq!(m, vec![Mention::File("src/a".into())]);
        let m2 = parse_mentions("@file 'b.rs'");
        assert_eq!(m2, vec![Mention::File("b.rs".into())]);
    }

    #[test]
    fn fence_lang_maps_common_extensions() {
        assert_eq!(fence_lang("a.rs"), "rust");
        assert_eq!(fence_lang("a.TSX"), "ts");
        assert_eq!(fence_lang("Makefile"), "");
    }

    #[test]
    fn file_block_wraps_with_fenced_lang() {
        let b = format_file_block("src/a.rs", "fn main() {}");
        assert!(b.contains("<file path=\"src/a.rs\">"));
        assert!(b.contains("```rust"));
        assert!(b.contains("fn main() {}"));
        assert!(b.contains("</file>"));
    }

    #[test]
    fn render_attached_is_identity_without_blocks() {
        assert_eq!(render_attached("hi", &[]), "hi");
        let r = render_attached("hi", &[format_file_block("a.rs", "x")]);
        assert!(r.contains("attached context") && r.starts_with("hi"));
    }
}

File: web/c5-ui/src/lib/agent/mentions.ts
/**
 * # Inline `@file` mention expansion
 *
 * Mirrors `crates/prompt/src/mentions.rs` (`agent_prompt::mentions`) so the Web
 * Composer and the Rust CLI expand `@file`-mentions into attached context blocks
 * with identical grammar and identical rendered format.
 *
 * - `@file <path>` (path is the next whitespace-delimited token) → injects the
 *   file body (`GET /api/file?path=<rel>`).
 *
 * Rendered as a context block appended after the user text. Parity is mandatory.
 */

/** A parsed mention. */
export type Mention = { kind: 'file'; path: string }

/**
 * Parse every `@file <path>` mention (deduped, first-seen order). Mirrors Rust
 * `parse_mentions` token-for-token:
 *
 * - Split by lines, then by whitespace.
 * - A standalone `@file` token consumes the next whitespace-delimited token as
 *   the path; surrounding `"`/`'` chars are stripped (any leading/trailing run).
 * - Unrecognized `@xxx` tokens are ignored.
 */
export function parseMentions(text: string): Mention[] {
  const out: Mention[] = []
  const hasFile = (p: string) => out.some((m) => m.kind === 'file' && m.path === p)
  for (const line of text.split('\n')) {
    const tokens = line.split(/\s+/).filter((t) => t.length > 0)
    for (let i = 0; i < tokens.length; i++) {
      const tok = tokens[i]
      if (tok.toLowerCase() === '@file') {
        const next = tokens[i + 1]
        if (next !== undefined) {
          // Strip any leading/trailing run of `"` or `'` (Rust trim_matches).
          const cleaned = next.replace(/^["']+|["']+$/g, '')
          if (!hasFile(cleaned)) out.push({ kind: 'file', path: cleaned })
          i += 1 // consume the path token
        }
      }
    }
  }
  return out
}

/**
 * Markdown fence language for an extension (mirrors Rust `fence_lang`).
 * Returns `''` for unknown extensions so the fence renders as a plain block.
 */
export function fenceLang(path: string): string {
  // Everything after the last `.`; a dotless path yields the whole name (also
  // unmatched → ''), matching Rust `rsplit('.').next()`.
  const dot = path.lastIndexOf('.')
  const ext = (dot === -1 ? path : path.slice(dot + 1)).toLowerCase()
  switch (ext) {
    case 'rs':
      return 'rust'
    case 'toml':
      return 'toml'
    case 'ts':
    case 'tsx':
      return 'ts'
    case 'js':
    case 'jsx':
    case 'mjs':
    case 'cjs':
      return 'js'
    case 'py':
      return 'python'
    case 'go':
      return 'go'
    case 'java':
      return 'java'
    case 'c':
    case 'h':
      return 'c'
    case 'cpp':
    case 'cc':
    case 'hpp':
      return 'cpp'
    case 'rb':
      return 'ruby'
    case 'sh':
    case 'bash':
    case 'zsh':
      return 'bash'
    case 'json':
      return 'json'
    case 'yaml':
    case 'yml':
      return 'yaml'
    case 'md':
      return 'md'
    case 'html':
      return 'html'
    case 'css':
      return 'css'
    default:
      return ''
  }
}

/** Render a single file context block (mirrors Rust `format_file_block`). */
export function formatFileBlock(path: string, content: string): string {
  return `<file path="${path}">\n\`\`\`${fenceLang(path)}\n${content}\n\`\`\`\n</file>`
}

/**
 * Append rendered context blocks after the user text. With no blocks the user
 * text is returned unchanged (mirrors Rust `render_attached`).
 */
export function renderAttached(userText: string, blocks: string[]): string {
  if (blocks.length === 0) return userText
  return `${userText}\n\n--- attached context ---\n\n${blocks.join('\n\n')}`
}

/** `/api/file` response shape (verified in `crates/server/src/lib.rs::read_file`). */
interface FileBody {
  path: string
  binary: boolean
  size: number
  truncated: boolean
  content: string | null
}

/** Context passed to {@link expandMentions}. */
export interface ExpandCtx {
  /** Authenticated GET helper (adds `?token=` / `&token=`). */
  apiGet: <T>(path: string) => Promise<T | null>
  /** Optional toast for skipped/failed attachments. Tolerant when absent. */
  say?: (text: string, level?: string) => void
  /** i18n translator（警告文案本地化必需，调用方必须传入）。 */
  t: (key: string, args?: Record<string, unknown>) => string
}

/**
 * Expand every `@file`-mention into attached context blocks and append them to
 * the user text. Tolerant: a failed/binary/missing file is skipped with an
 * optional `say` warning rather than aborting the send.
 *
 * Returns the original text unchanged when there are no mentions.
 */
export async function expandMentions(text: string, ctx: ExpandCtx): Promise<string> {
  const mentions = parseMentions(text)
  if (mentions.length === 0) return text
  const blocks: string[] = []
  for (const m of mentions) {
    const data = await ctx.apiGet<FileBody>(`/api/file?path=${encodeURIComponent(m.path)}`)
    if (data && data.content !== null && !data.binary) {
      blocks.push(formatFileBlock(m.path, data.content))
    } else {
      ctx.say?.(ctx.t('mentions.attach_failed', { path: m.path }), 'warning')
    }
  }
  return renderAttached(text, blocks)
}
